src/py/flwr/client/process/process.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower background ClientApp."""
+
+from logging import DEBUG, ERROR, INFO
+from typing import Tuple
+
+import grpc
+
+from flwr.client.client_app import ClientApp
+from flwr.common import Context, Message
+from flwr.common.grpc import create_channel
+from flwr.common.logger import log
+from flwr.common.serde import (
+    context_from_proto,
+    context_to_proto,
+    message_from_proto,
+    message_to_proto,
+    run_from_proto,
+)
+from flwr.common.typing import Run
+
+# pylint: disable=E0611
+from flwr.proto.clientappio_pb2 import (
+    PullClientAppInputsRequest,
+    PullClientAppInputsResponse,
+    PushClientAppOutputsRequest,
+    PushClientAppOutputsResponse,
+)
+from flwr.proto.clientappio_pb2_grpc import ClientAppIoStub
+
+from .utils import _get_load_client_app_fn
+
+
+def _run_background_client(  # pylint: disable=R0914
+    address: str,
+    token: int,
+) -> None:
+    """Run background Flower ClientApp process.
+
+    Parameters
+    ----------
+    address : str
+        Address of SuperNode
+    token : int
+        Unique SuperNode token for ClientApp-SuperNode authentication
+    """
+
+    def on_channel_state_change(channel_connectivity: str) -> None:
+        """Log channel connectivity."""
+        log(DEBUG, channel_connectivity)
```

Comment:
Why is this function in-line and not top level?

Reply:
Good question. There isn't a strong reason for this apart from possibly changing the callback behaviour in the future. For context, I replicated the code in `cli/run/run.py` where `on_channel_state_change` is in-line. I see that in `grpc_rere_client/connection.py that it is top level. Wdyt @danieljanes

Reply:
Moved to top-level.